src/impl_methods.rs

```diff
@@ -2444,6 +2444,46 @@ where
         }
     }
 
+    /// Remove the `index`th elements along `axis` and shift down elements from higher indexes.
+    ///
+    /// Decreases the length of `axis` by one.
+    ///
+    /// ***Panics** if `axis` or `index` is out of bounds.
+    pub fn shift_remove_index(&mut self, axis: Axis, index: usize)
+    where
+        S: DataOwned + DataMut,
+    {
+        // TODO: It's possible to choose to shift the elment to the front or the back here,
+        // whichever is closer.
+        let (_, mut tail) = self.view_mut().split_at(axis, index);
+        // shift elements to the back
+        // use swap to keep all elements initialized (as required by owned storage)
+        Zip::from(tail.lanes_mut(axis)).for_each(|mut lane| {
+            let mut lane_iter = lane.iter_mut();
+            let mut dst = if let Some(dst) = lane_iter.next() { dst } else { return };
+
+            // Logically we do a circular swap here, all elements in a chain
+            // Using MaybeUninit to avoid unecessary writes in the safe swap solution
+            //
+            //  for elt in rest.iter_mut() {
+            //      std::mem::swap(dst, elt);
+            //      dst = elt;
+            //  }
+            //
+            let mut slot = MaybeUninit::<A>::uninit();
+            unsafe {
+                slot.as_mut_ptr().copy_from_nonoverlapping(dst, 1);
+                for elt in lane_iter {
```

Comment:
If `lane_iter.next()` panics, then an element will be dropped twice, and, less importantly, the element in `slot` won't be dropped. I don't think the current implementation can panic, but this still makes me uncomfortable. I'd feel more comfortable with getting a raw pointer to the start of the lane, the stride for the axis, and then `.offset()`ting the pointer by the stride on each iteration, because this would be easier to audit for potential panics.

Reply:
Yes I know this is a no-panic section and it should be marked. I was comfortable apparently, but I had a thought to maybe leave an "abort on panic" guard here to make it entirely visible.

As for a simpler iteration, in that case we should implement a special 1D iterator that does just that (even though Zip is already doing about the same job). I don't want to write out a raw pointer loop, ndarray probably needs a bit more use of higher level abstractions, not less 🙂